snippet_parser/core.py

```diff
@@ -191,94 +193,120 @@ def extract(self, wikitext):
-            else:
-                # Throw away the actual template, we don't need it.
-                for marker in tree.cssselect('.' + CITATION_NEEDED_MARKER_CLASS):
-                    lxml_utils.remove_element(marker)
-
-                # Keep only snippet root top-level elements within the body
-                # that have any text content (we may have created empty elements
-                # above during cleanup). This is not great as any content within,
-                # say, <blockquote> gets removed entirely, but it's good enough
-                # in most cases.
-                snippet_roots = [
-                    self._make_snippet_root(*(
-                        e for e in tree.cssselect(
-                            'body > ' + ', '.join(_SNIPPET_ROOT_TAGS))
-                        if e.text_content() and not e.text_content().isspace()))
-                ]
-
-            snippets_in_section = set()
-            for sr in snippet_roots:
-                # Some last-minute cleanup to shrink the snippet some more.
-                # Remove links and attributes, but make sure to keep the
-                # class in our marker elements, and that there is no space
-                # before it (which we need for the UI).
-                lxml.etree.strip_tags(sr, 'a')
-                markers_in_snippet = sr.cssselect(
-                    '.' + CITATION_NEEDED_MARKER_CLASS)
-                lxml.etree.strip_attributes(sr, 'id', 'class', 'style')
-                sr.attrib['class'] = SNIPPET_WRAPPER_CLASS
-                for marker in markers_in_snippet:
-                    marker.attrib['class'] = CITATION_NEEDED_MARKER_CLASS
-                    lxml_utils.strip_space_before_element(marker)
-
-                length = len(sr.text_content().strip())
-                self.stats.snippet_lengths[length] += 1
-                if minlen < length < maxlen:
-                    snippet = d(lxml.html.tostring(
-                        sr, encoding = 'utf-8', method = 'html')).strip()
-                    snippets_in_section.add(snippet)
-
-            sectitle = ''
-            if i != 0:
-                # Re-parse the section title because fast_parse is
-                # configured to ignore style tags (see above and 
-                # https://github.com/earwig/mwparserfromhell/issues/40),
-                # but we do want to remove them now with strip_code().
-                sectitle = mwparserfromhell.parse(
-                    str(section.get(0).title).strip()).strip_code()
+            snippets_in_section = self._parse_snippets(
+                html, CITATION_NEEDED_MARKER_CLASS, minlen, maxlen)
+            if snippets_in_section is None: continue
+            sectitle = self._parse_section_title(i, section)
+            snippets.append([sectitle, list(snippets_in_section)])
+        return snippets
+
+    def extract_from_sentences(self, wikitext, sentences):
```

Comment:
Please add some unit tests to snippet_parser/core_test.py to check that this is working. A couple of examples of things you might want to test for:

- Passing a sentence that is not found on the Wikitext
- A "happy" case where a sentence turns into a snippet just fine
- Repeated sentences
- Sentences that are larger than the max size for a snippet

You can run the tests with `python snippet_parser/core_test.py` by the way.

Reply:
Hi @eggpi, I've updated the code :)

Unittest library is new to me, so I was trying to read your code and some other resource to understand it. The following is what I understand. Please correct me if I'm wrong.
- You use mock.Mock() so you can fabricate html that returns from parse API.
- In most test cases, you test extracting snippets from different html cases.
- In `test_drop_ref_groups`, it's a bit different. You test if the cleaned wikitext is right using `call_args`

Based on the knowledge I've learned, I was thinking how to write the test code. There are some different behaviors: 
- `_extract_from_sentences` has two args, `sentence` list and `wikitext`.
- If the same sentence is found in the wikitext, the wikitext will be inserted a html marker. 
- Then, the wikitext is converted to html, and extract snippets from html. (same as `extract`)

In the end, I found I need to call the function `_do_extract_citationdetective` twice, the first one is to get the modified wikitext (call_args) and the second one is to get the return snippets by passing the modified wikitext. 

I feel there would be a better way to do this and maybe you can clear something up for me. I really appreciate it :)

Reply:
> In `test_drop_ref_groups`, it's a bit different. You test if the cleaned wikitext in `call_args` is right.

Yep, ultimately I'm testing that we remove the groups from `<ref>` tags **before** sending them to the API.

> In the end, I found I need to call the function `_do_extract_citationdetective` twice (you can see it in my code), the first one is to get the modified wikitext (call_args) and the second one is to get the return snippets by passing the modified wikitext.

Wait, this part sounds wrong. As you noticed, `_do_extract` is just a convenience function that returns both the intermediate marked wikitext that we send to the API, and the final results after extracting snippets from the HTML. With one call to `_do_extract`, we should be able to validate either or both of those, if we wanted to -- though there are no tests that use both currently.

Likewise, I would expect the new tests to only require one call to `extract_sentences`: before we call it, we set up all the fake values we need with mocks, then we call the convenience `_do_extract_citationdetective`, which returns everything we need to check that things went right.

Why would we want to call `extract_sentences` twice? If that's not how the method is supposed to be called in "real" code, we should not be doing that in tests either. After all, a test is supposed to simulate real uses of the code and check that it behaves well under realistic conditions.

Reply:
> Why would we want to call `extract_sentences` twice? If that's not how the method is supposed to be called in "real" code, we should not be doing that in tests either. After all, a test is supposed to simulate real uses of the code and check that it behaves well under realistic conditions.

Because I wanted to know both "the intermediate marked wikitext" and "the final snippets extracted from the HTML". For later one, it depends on the intermediate marked wikitext. But yeah, now I realize to call `extract_sentences` twice doesn't make sense because that's not how we do in a real use case.

I think I should divide each test case into two test cases. For example, if I want to test a case where a sentence turns into a snippet is fine, one test case to check the intermediate marked wikitext is right by setting up a fake sentence and a wikitext; another test case to check the final snippets is right by setting up a fake intermediate marked HTML. Does it sound right?

Reply:
Hm, wait, not quite. Apologies if I'm misunderstanding, I have a cold so I'm not totally functional. We can chat tomorrow if I'm missing anything.

So if you look at how `_do_extract_` works, it:

1. Sets up the mock API to [return a fake HTML](https://github.com/eggpi/citationhunt/blob/master/snippet_parser/core_test.py#L33-L36), when `extract` does the HTML conversion.
2. Calls `extract` with the wikitext we're testing for. Here, `extract` runs normally and produces a result, only using the fake data we provided.
3. [Returns to the caller](https://github.com/eggpi/citationhunt/blob/master/snippet_parser/core_test.py#L38-L42) a tuple `(marked wikitext, [section title, list of snippets])`.

That is, the use of mock allows us to both return a fake HTML when the marked wikitext is passed to the fake API, _and_ retrieve what the marked wikitext was, after `extract` has returned.

So even if you want to test the intermediate marked wikitext, one call per test case should be enough. I expect the same would be true for `_do_extract_sentences`: we call it once per test and each time it returns all the data we need to validate that things worked, in the scenario we're simulating.

Also - when testing, usually we want to verify only the externally observable behaviors of the method. In other words, usually in tests we don't actually want to verify the marked wikitext, since in real code, the callers of the method never get to see it -- they only get the final HTML. That is the reason why most tests in `core_test.py` actually ignore the marked wikitext returned from `_do_extract`, except for `test_drop_ref_groups`.

Does that clarify things?